contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -4453,3 +4608,691 @@ rename_tsql_db(char *old_db_name, char *new_db_name)
 	if (!xactStarted)
 		CommitTransactionCommand();
 }
+
+/*
+ * partition_function_exists_with_id
+ * 		Helper function to generate new ids for partition function,
+ * 		it check if provided id is already in use.
+ */
+static bool
+partition_function_exists_with_id(int32 id)
+{
+	Relation	rel;
+	HeapTuple	tuple;
+	SysScanDesc	scan;
+	ScanKeyData	scanKey;
+	bool		partition_function_exists = false;
+	/* open the relation */
+	rel = table_open(get_bbf_partition_function_oid(), AccessShareLock);
+	
+	ScanKeyInit(&scanKey,
+					Anum_bbf_partition_function_id,
+					BTEqualStrategyNumber, F_INT2EQ,
+					Int32GetDatum(id));
+
+	/* scan using index */
+	scan = systable_beginscan(rel,
+			get_bbf_partition_function_id_idx_oid(),
+			false, NULL, 1, &scanKey);
+	
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+		partition_function_exists = true;
+
+	systable_endscan(scan);
+	table_close(rel, AccessShareLock);
+	return partition_function_exists;
+}
+
+/*
+ * get_available_partition_function_id
+ * 		Generates new ids for partition scheme.
+ * 		1. To guard against race conditions for IDs, we check if the 
+ * 		   ID generated by sequence is used by an existing partition function.
+ * 		2. To guard rare case where all possible sequence values have been exhausted
+ * 		   and the sequence wraps around, we will loop through the entire range of
+ * 		   sequence values and on loop completion, we should bail out.
+ */
+int32
+get_available_partition_function_id(void)
+{
+	int32		id;
+	int32		start = 0;
+	bool		loop_completed = false;
+
+	do
+	{
+		id = nextval_internal(get_bbf_partition_function_seq_oid(), false);
+		if (start == 0)
+			start = id;
+		else if (start == id)
+		{
+			loop_completed = true;
+			break;
+		}
+	} while (partition_function_exists_with_id(id));
+
+	if (loop_completed)
```

Comment:
This is a race condition, you can try by attach gdb on one process here and in another process you will fetch the same id since partition_function_exists_with_id will return false (process 1 has not been committed yet)

Reply:
I think we can use LOCKTAG to avoid

Reply:
`nextval_internal` itself handles this, internally it uses lock on sequence to generate unique number for each function call and even if it is called during parallel mode. Also, if the transaction is not committed the sequence value will be counted as committed.